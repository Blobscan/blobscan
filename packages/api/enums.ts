export { BlobStorage, Category, Rollup } from "@blobscan/db/prisma/enums";
