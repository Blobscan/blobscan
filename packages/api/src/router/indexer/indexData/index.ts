import type { Blob, Transaction, OmittableFields, Block } from "@blobscan/db";

import { tracer } from "../../../instrumentation";
import { jwtAuthedProcedure } from "../../../procedures";
import {
  calculateBlobGasPrice,
  calculateBlobSize,
  getEIP2028CalldataGas,
} from "../../../utils/ethereum";
import { INDEXER_PATH } from "../constants";
import type { IndexDataInput } from "./schema";
import { indexDataOutputSchema } from "./schema";
import { indexDataInputSchema } from "./schema";

function createDBTransactions({
  blobs,
  block,
  transactions,
}: IndexDataInput): Omit<Transaction, OmittableFields>[] {
  return transactions.map<Omit<Transaction, OmittableFields>>(
    ({ blockNumber, from, gasPrice, hash, maxFeePerBlobGas, to }) => {
      const txBlob: IndexDataInput["blobs"][0] | undefined = blobs.find(
        (b) => b.txHash === hash
      );

      if (!txBlob) {
        throw new Error(`Blob for transaction ${hash} not found`);
      }

      return {
        blockNumber,
        hash,
        fromId: from,
        toId: to,
        gasPrice,
        blobGasPrice: calculateBlobGasPrice(block.excessBlobGas),
        maxFeePerBlobGas,
        blobAsCalldataGasUsed: getEIP2028CalldataGas(txBlob.data),
      };
    }
  );
}

function createDBBlock(
  {
    block: { blobGasUsed, excessBlobGas, hash, number, slot, timestamp },
  }: IndexDataInput,
  dbTxs: { blobAsCalldataGasUsed: bigint }[]
): Omit<Block, OmittableFields> {
  const blobAsCalldataGasUsed = dbTxs.reduce(
    (acc, tx) => acc + tx.blobAsCalldataGasUsed,
    BigInt(0)
  );

  return {
    number,
    hash,
    timestamp: new Date(timestamp * 1000),
    slot,
    blobGasUsed,
    blobGasPrice: calculateBlobGasPrice(excessBlobGas),
    excessBlobGas,
    blobAsCalldataGasUsed,
  };
}

function createDBBlobs({
  blobs,
  block,
}: IndexDataInput): Omit<Blob, OmittableFields>[] {
  return blobs.map<Omit<Blob, OmittableFields>>((blob) => ({
    versionedHash: blob.versionedHash,
    commitment: blob.commitment,
    size: calculateBlobSize(blob.data),
    firstBlockNumber: block.number,
  }));
}

export const indexData = jwtAuthedProcedure
  .meta({
    openapi: {
      method: "PUT",
      path: `${INDEXER_PATH}/block-txs-blobs`,
      tags: ["indexer"],
      summary: "Index data in the database",
      protect: true,
    },
  })
  .input(indexDataInputSchema)
  .output(indexDataOutputSchema)
  .mutation(async ({ ctx: { prisma, blobStorageManager }, input }) => {
    const operations = [];

    // 1. Upload blobs' data to storages

    const newBlobs = await prisma.blob.filterNewBlobs(input.blobs);

    const blobUploadResults = await tracer.startActiveSpan(
      "blobs-upload",
      async (blobsUploadSpan) => {
        const result = (
          await Promise.all(
            newBlobs.map(async (b) =>
              blobStorageManager.storeBlob(b).then<{
                uploadRes: Awaited<
                  ReturnType<typeof blobStorageManager.storeBlob>
                >;
                blob: Awaited<
                  ReturnType<typeof prisma.blob.filterNewBlobs>
                >[number];
              }>((uploadRes) => ({
                blob: b,
                uploadRes,
              }))
            )
          )
        ).flat();

        blobsUploadSpan.end();

        return result;
      }
    );

    blobUploadResults.forEach(({ uploadRes: { errors }, blob }) => {
      if (errors.length) {
        const errorMsgs = errors.map((e) => `${e.storage}: ${e.error}`);

        console.warn(
          `Couldn't upload blob ${
            blob.versionedHash
          } to some of the storages: ${errorMsgs.join(", ")}`
        );
      }
    });

    const dbBlobStorageRefs = blobUploadResults.flatMap(
      ({ uploadRes: { references }, blob }) =>
        references.map((ref) => ({
          blobHash: blob.versionedHash,
          blobStorage: ref.storage,
          dataReference: ref.reference,
        }))
    );

    // 2. Prepare address, block, transaction and blob insertions

    // TODO: Create an upsert extension that set the `insertedAt` and the `updatedAt` field
    const now = new Date();

    const dbTxs = createDBTransactions(input);
    const dbBlock = createDBBlock(input, dbTxs);
    const dbBlobs = createDBBlobs(input);

    operations.push(
      prisma.block.upsert({
        where: { hash: input.block.hash },
        create: {
          ...dbBlock,
          insertedAt: now,
          updatedAt: now,
        },
        update: {
          ...dbBlock,
          updatedAt: now,
        },
      })
    );
    operations.push(
      prisma.address.upsertAddressesFromTransactions(input.transactions)
    );
    operations.push(prisma.transaction.upsertMany(dbTxs));
    operations.push(prisma.blob.upsertMany(dbBlobs));

    if (dbBlobStorageRefs.length) {
      operations.push(
        prisma.blobDataStorageReference.upsertMany(dbBlobStorageRefs)
      );
    }

    operations.push(
      prisma.blobsOnTransactions.createMany({
        data: input.blobs.map((blob) => ({
          blobHash: blob.versionedHash,
          txHash: blob.txHash,
          index: blob.index,
        })),
        skipDuplicates: true,
      })
    );

    // 3. Execute all database operations in a single transaction
    await prisma.$transaction(operations);
  });
