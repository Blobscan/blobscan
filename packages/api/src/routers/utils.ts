export const GAS_PER_BLOB = 131_072; // 2 ** 17
export const BLOB_GAS_LIMIT_PER_BLOCK = 786_432;

function isValidInt(number: number): boolean {
  const minInt = -2147_483_648;
  const maxInt = 2147_483_647;

  return number >= minInt && number <= maxInt;
}
/**
 *
 * Checks if the given string is a number
 */
export function isBlockNumber(number: string): boolean {
  const number_ = Number(number);

  return !isNaN(number_) && isValidInt(number_);
}
