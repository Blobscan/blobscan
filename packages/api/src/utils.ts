import { CHAIN_ID } from "./env";

const ADDRESS_REGEX = /^0x[0-9a-fA-F]{40}$/;
const HASH_REGEX = /^0x[A-Fa-f0-9]{64}$/;
const COMMITMENT_REGEX = /^0x[A-Fa-f0-9]{96}$/;

function isValidInt(number: number): boolean {
  const minInt = -2147_483_648;
  const maxInt = 2147_483_647;

  return number >= minInt && number <= maxInt;
}

/**
 *
 * Checks if the given string is an address
 */
export function isAddress(address: string): boolean {
  return ADDRESS_REGEX.test(address);
}

/**
 *
 * Checks if the given string is a hash
 */
export function isHash(hash: string): boolean {
  return HASH_REGEX.test(hash);
}

/**
 *
 * Checks if the given string is a KZG commitment
 */
export function isCommitment(commitment: string): boolean {
  return COMMITMENT_REGEX.test(commitment);
}

/**
 *
 * Checks if the given string is a number
 */
export function isBlockNumber(number: string): boolean {
  const number_ = Number(number);

  return !isNaN(number_) && isValidInt(number_);
}

/**
 *
 * Builds a Google Storage URI for the given hash
 *
 */
export function buildGoogleStorageUri(hash: string): string {
  return `${CHAIN_ID}/${hash.slice(2, 4)}/${hash.slice(4, 6)}/${hash.slice(
    6,
    8,
  )}/${hash.slice(2)}.blob`;
}
