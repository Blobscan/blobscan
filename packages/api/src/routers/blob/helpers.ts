import type {
  BlobscanPrismaClient,
  EthUsdPrice,
  ExtendedBlobDataStorageReferenceSelect,
  Prisma,
} from "@blobscan/db";
import { EthUsdPriceModel } from "@blobscan/db/prisma/zod";
import { z } from "@blobscan/zod";

import type {
  ExpandedBlob,
  ExpandedBlock,
  ExpandedTransaction,
  Expands,
} from "../../middlewares/withExpands";
import type { Prettify } from "../../types";
import { isFullyDefined } from "../../utils";
import {
  normalizePrismaBlobDataStorageReferencesFields,
  normalizePrismaTransactionFields,
} from "../../utils";
import {
  baseBlobSchema,
  baseBlockSchema,
  baseTransactionSchema,
  prismaBlobOnTransactionSchema,
} from "../../zod-schemas";

const blobDataStorageReferenceSelect = {
  blobStorage: true,
  url: true,
} satisfies ExtendedBlobDataStorageReferenceSelect;

export const extendedBlobDataStorageReferenceArgs: Prisma.Args<
  BlobscanPrismaClient["blobDataStorageReference"],
  "findFirstOrThrow"
> = {
  select: blobDataStorageReferenceSelect,
  orderBy: {
    blobStorage: "asc",
  },
};

export const baseBlobSelect = {
  commitment: true,
  proof: true,
  size: true,
  usageSize: true,
  versionedHash: true,
} satisfies Prisma.BlobSelect;

export const baseBlobOnTransactionSelect = {
  blobHash: true,
  blockHash: true,
  blockNumber: true,
  blockTimestamp: true,
  index: true,
  txHash: true,
  txIndex: true,
} satisfies Prisma.BlobsOnTransactionsSelect;

export type ExtendedBlobDataStorageReference = NonNullable<
  Prisma.Result<
    BlobscanPrismaClient["blobDataStorageReference"],
    { select: typeof blobDataStorageReferenceSelect },
    "findFirst"
  >
>;
type PrismaBlob = Prisma.BlobGetPayload<{
  select: typeof baseBlobSelect;
}> & {
  dataStorageReferences: ExtendedBlobDataStorageReference[];
};

type PrismaBlobOnTransaction = Prisma.BlobsOnTransactionsGetPayload<{
  select: typeof baseBlobOnTransactionSelect;
}>;

export type CompletedPrismaBlob = Prettify<
  PrismaBlob & {
    transactions: Prettify<
      PrismaBlobOnTransaction & {
        block?: ExpandedBlock;
        transaction?: ExpandedTransaction & {
          block: { blobGasPrice: ExpandedBlock["blobGasPrice"] };
        };
      }
    >[];
  }
>;

export type CompletedPrismaBlobOnTransaction = Prettify<
  PrismaBlobOnTransaction & {
    blob: {
      dataStorageReferences: ExtendedBlobDataStorageReference[];
    } & ExpandedBlob;
    block?: ExpandedBlock;
    transaction?: ExpandedTransaction & {
      block: { blobGasPrice: ExpandedBlock["blobGasPrice"] };
    };
  }
>;
export function createBlobSelect(expands: Expands) {
  const blockExpand = expands.block ? { block: expands.block } : {};
  const txExpand = expands.transaction
    ? {
        transaction: {
          select: {
            ...expands.transaction.select,
            block: {
              select: {
                blobGasPrice: true,
              },
            },
          },
        },
      }
    : {};

  return {
    ...baseBlobSelect,
    dataStorageReferences: extendedBlobDataStorageReferenceArgs,
    transactions: {
      select: {
        ...baseBlobOnTransactionSelect,
        ...blockExpand,
        ...txExpand,
      },
      orderBy: {
        blockNumber: "desc",
      },
    },
  } satisfies Prisma.BlobSelect;
}

export function createBlobsOnTransactionsSelect(expands: Expands) {
  const blockExpand = expands.block ? { block: expands.block } : {};
  const txExpand = expands.transaction
    ? {
        transaction: {
          select: {
            ...expands.transaction.select,
            block: {
              select: {
                blobGasPrice: true,
              },
            },
          },
        },
      }
    : {};

  return {
    ...baseBlobOnTransactionSelect,
    blob: {
      select: {
        ...baseBlobSelect,
        dataStorageReferences: extendedBlobDataStorageReferenceArgs,
      },
    },
    ...txExpand,
    ...blockExpand,
  } satisfies Prisma.BlobsOnTransactionsSelect;
}

const responseBlobOnTxSchema = prismaBlobOnTransactionSchema
  .omit({ blobHash: true })
  .required({
    blockHash: true,
    blockNumber: true,
    blockTimestamp: true,
    index: true,
    txHash: true,
    txIndex: true,
  });

export const responseBlobSchema = baseBlobSchema.extend({
  transactions: z.array(
    responseBlobOnTxSchema
      .merge(
        baseTransactionSchema
          .omit({
            hash: true,
            blockNumber: true,
            blockTimestamp: true,
            blockHash: true,
          })
          .partial()
      )
      .extend({
        block: baseBlockSchema
          .omit({
            hash: true,
            number: true,
            timestamp: true,
          })
          .partial({
            blobGasBaseUsdFee: true,
            blobGasUsdPrice: true,
          })
          .optional(),
        ethUsdPrice: EthUsdPriceModel.shape.price.optional(),
      })
  ),
});

export const responseBlobOnTransactionSchema = responseBlobOnTxSchema
  .merge(baseBlobSchema)
  .extend({
    block: baseBlockSchema
      .omit({
        hash: true,
        number: true,
        timestamp: true,
      })
      .optional(),
    transaction: baseTransactionSchema
      .omit({
        hash: true,
        blockNumber: true,
        blockTimestamp: true,
        blockHash: true,
      })
      .optional(),
  });

export type ResponseBlob = z.input<typeof responseBlobSchema>;
export type ResponseBlobOnTransaction = z.input<
  typeof responseBlobOnTransactionSchema
>;

export function toResponseBlob(
  {
    dataStorageReferences,
    transactions: blobOnTxs,
    ...prismaBlob
  }: CompletedPrismaBlob,
  ethUsdPrices: EthUsdPrice["price"][]
): ResponseBlob {
  const responseBlob: ResponseBlob = {
    ...prismaBlob,
    dataStorageReferences: normalizePrismaBlobDataStorageReferencesFields(
      dataStorageReferences
    ),
    transactions: blobOnTxs.map(
      (
        {
          block: { computeUsdFields, ...restBlock } = {},
          transaction: prismaTx = {},
          ...restBlobOnTx
        },
        i
      ) => {
        const txEthUsdPrice = ethUsdPrices[i];

        const blockUsdFields =
          txEthUsdPrice && computeUsdFields
            ? computeUsdFields(txEthUsdPrice)
            : undefined;
        const txBlock = {
          ...restBlock,
          ...(blockUsdFields ?? {}),
        };

        let responseBlobOnTx: ResponseBlob["transactions"][number] = {
          ...restBlobOnTx,
          ...txBlock,
          ethUsdPrice: txEthUsdPrice?.toNumber(),
        };

        if (isFullyDefined(prismaTx)) {
          const {
            block,
            decodedFields,
            from,
            fromId,
            toId,
            computeBlobGasBaseFee,
            computeUsdFields,
            ...restPrismaTx
          } = prismaTx;

          responseBlobOnTx = {
            ...responseBlobOnTx,
            ...restPrismaTx,
            ...normalizePrismaTransactionFields({
              decodedFields,
              from,
              fromId,
              toId,
            }),
            blobGasBaseFee: computeBlobGasBaseFee(block.blobGasPrice),
            ...(txEthUsdPrice
              ? computeUsdFields({
                  blobGasPrice: block.blobGasPrice,
                  ethUsdPrice: txEthUsdPrice,
                })
              : {}),
          };
        }

        return responseBlobOnTx;
      }
    ),
  };

  return responseBlob;
}

export function toResponseBlobOnTransaction(
  prismaBlobOnTransaction: CompletedPrismaBlobOnTransaction,
  ethUsdPrice?: EthUsdPrice["price"]
): ResponseBlobOnTransaction {
  const {
    blobHash,
    blob: { dataStorageReferences, ...blob },
    transaction,
    block,
    ...restBlobOnTransaction
  } = prismaBlobOnTransaction;

  const responseBlobOnTx: ResponseBlobOnTransaction = {
    ...blob,
    versionedHash: blobHash,
    dataStorageReferences: normalizePrismaBlobDataStorageReferencesFields(
      dataStorageReferences
    ),
    ...restBlobOnTransaction,
  };

  if (transaction) {
    const {
      block: txBlock,
      decodedFields,
      from,
      fromId,
      toId,
      computeBlobGasBaseFee,
      computeUsdFields,
      ...restTx
    } = transaction;

    responseBlobOnTx.transaction = {
      ...restTx,
      blobGasPrice: txBlock.blobGasPrice,
      ...normalizePrismaTransactionFields({
        decodedFields,
        from,
        fromId,
        toId,
      }),
      blobGasBaseFee: computeBlobGasBaseFee(txBlock.blobGasPrice),
      ...(ethUsdPrice
        ? computeUsdFields({
            blobGasPrice: txBlock.blobGasPrice,
            ethUsdPrice,
          })
        : {}),
    };
  }

  if (block) {
    const { computeUsdFields, ...restBlock } = block;
    responseBlobOnTx.block = {
      ...restBlock,
      ...(ethUsdPrice ? computeUsdFields(ethUsdPrice) : {}),
    };
  }

  return responseBlobOnTx;
}
