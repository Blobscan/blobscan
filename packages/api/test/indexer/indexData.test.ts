import { TRPCError } from "@trpc/server";
import type { SpyInstance } from "vitest";
import {
  afterAll,
  beforeAll,
  beforeEach,
  describe,
  expect,
  it,
  vi,
} from "vitest";

import { Category, Rollup } from "@blobscan/db/prisma/enums";
import { getNetworkBlobConfigBySlot } from "@blobscan/network-blob-config";
import { ADDRESS_TO_ROLLUP_MAPPINGS } from "@blobscan/rollups";
import { env, omitDBTimestampFields, testValidError } from "@blobscan/test";

import type { TRPCContext } from "../../src";
import { calculateBlobGasPrice } from "../../src/routers/indexer/indexData.utils";
import { createTestContext } from "../helpers";
import { createIndexerCaller } from "./caller";
import type { IndexerCaller } from "./caller";
import {
  INPUT,
  INPUT_WITH_DUPLICATED_BLOBS,
  ROLLUP_BLOB_TRANSACTION_INPUT,
} from "./fixtures";

describe("indexData", () => {
  let nonAuthorizedCaller: IndexerCaller;
  let authorizedCaller: IndexerCaller;
  let noPropagatorCaller: IndexerCaller;
  let authorizedCtx: TRPCContext;

  beforeAll(async () => {
    const ctx = await createTestContext({
      withBlobPropagator: true,
    });

    authorizedCtx = await createTestContext({
      apiClient: "indexer",
      withBlobPropagator: true,
    });

    nonAuthorizedCaller = createIndexerCaller(ctx);
    authorizedCaller = createIndexerCaller(authorizedCtx);
    noPropagatorCaller = createIndexerCaller(
      await createTestContext({
        apiClient: "indexer",
        withBlobPropagator: false,
      })
    );
  });

  afterAll(async () => {
    vi.clearAllMocks();
    vi.unstubAllEnvs();
  });

  describe("when authorized", () => {
    beforeEach(async () => {
      await authorizedCaller.indexData(INPUT);
    });

    it("should index a block correctly", async () => {
      const indexedBlock = await authorizedCtx.prisma.block
        .findUnique({
          where: {
            hash: INPUT.block.hash,
          },
        })
        .then((res) => (res ? omitDBTimestampFields(res) : res));

      const {
        blobGasPrice,
        blobAsCalldataGasUsed: _,
        ...remainingParams
      } = indexedBlock ?? {};
      // const expectedBlobAsCalldataGasUsed = INPUT.blobs.reduce(
      //   (acc, b) => acc + getEIP2028CalldataGas(b.data),
      //   0
      // );
      const config = getNetworkBlobConfigBySlot(env.CHAIN_ID, INPUT.block.slot);
      const expectedBlobGasPrice = calculateBlobGasPrice(
        BigInt(INPUT.block.excessBlobGas),
        config
      );

      // TODO: Fix this test
      // expect(blobAsCalldataGasUsed).toBe(expectedBlobAsCalldataGasUsed);
      expect(blobGasPrice?.toString(), "Blob gas price mismatch").toBe(
        expectedBlobGasPrice.toString()
      );
      expect(remainingParams).toMatchInlineSnapshot(`
          {
            "blobGasBaseFee": "10000",
            "blobGasUsed": "10000",
            "computeUsdFields": [Function],
            "excessBlobGas": "5000",
            "hash": "blockHash2010",
            "number": 2010,
            "slot": 130,
            "timestamp": 2023-09-01T13:50:21.000Z,
            Symbol(nodejs.util.inspect.custom): [Function],
          }
        `);
    });

    describe("when indexing transactions", () => {
      const expectedRollup = Rollup.ARBITRUM;

      beforeAll(() => {
        ADDRESS_TO_ROLLUP_MAPPINGS.set(
          env.CHAIN_ID,
          new Map([
            ...ROLLUP_BLOB_TRANSACTION_INPUT.transactions.map(
              (tx) => [tx.from, expectedRollup] as [string, Rollup]
            ),
          ])
        );
      });

      afterAll(() => {
        ADDRESS_TO_ROLLUP_MAPPINGS.delete(env.CHAIN_ID);
      });

      it("should index transactions correctly", async () => {
        const indexedTxs = await authorizedCtx.prisma.transaction
          .findMany({
            where: {
              blockHash: INPUT.block.hash,
            },
            orderBy: [
              {
                blockNumber: "asc",
              },
              {
                index: "asc",
              },
            ],
          })
          .then((r) => r.map(omitDBTimestampFields));
        // const expectedBlobAsCalldataGasUsed = INPUT.transactions.map((tx) =>
        //   INPUT.blobs
        //     .filter((b) => b.txHash === tx.hash)
        //     .reduce((acc, b) => acc + getEIP2028CalldataGas(b.data), 0)
        // );
        // const blobAsCalldataGasUsed = indexedTxs.map(
        //   (tx) => tx.blobAsCalldataGasUsed
        // );
        const remainingParams = indexedTxs.map(
          ({
            blobAsCalldataGasUsed: _,
            computeBlobGasBaseFee: __,
            computeUsdFields: ___,
            blobAsCalldataGasFee: ____,
            blobGasMaxFee: _____,
            ...remainingParams
          }) => remainingParams
        );

        // TODO: Fix this test
        // expect(
        //   blobAsCalldataGasUsed,
        //   "Transactions' blob as calldata gas used mismatch"
        // ).toEqual(expectedBlobAsCalldataGasUsed);
        expect(remainingParams).toMatchInlineSnapshot(`
            [
              {
                "blobGasUsed": "131072",
                "blockHash": "blockHash2010",
                "blockNumber": 2010,
                "blockTimestamp": 2023-09-01T13:50:21.000Z,
                "decodedFields": {},
                "fromId": "address9",
                "gasPrice": "10000",
                "hash": "txHash999",
                "index": 0,
                "maxFeePerBlobGas": "1800",
                "toId": "address10",
                Symbol(nodejs.util.inspect.custom): [Function],
              },
              {
                "blobGasUsed": "262144",
                "blockHash": "blockHash2010",
                "blockNumber": 2010,
                "blockTimestamp": 2023-09-01T13:50:21.000Z,
                "decodedFields": {},
                "fromId": "address7",
                "gasPrice": "3000000",
                "hash": "txHash1000",
                "index": 1,
                "maxFeePerBlobGas": "20000",
                "toId": "address2",
                Symbol(nodejs.util.inspect.custom): [Function],
              },
            ]
          `);
      });

      it("should identify rollup transactions correctly", async () => {
        await authorizedCaller.indexData(ROLLUP_BLOB_TRANSACTION_INPUT);

        const indexedTxHashesAndRollups = await authorizedCtx.prisma.transaction
          .findMany({
            select: {
              hash: true,
              from: {
                select: {
                  rollup: true,
                },
              },
            },
            where: {
              blockHash: ROLLUP_BLOB_TRANSACTION_INPUT.block.hash,
            },
          })
          .then((r) => r.map((tx) => [tx.hash, tx.from.rollup]));

        const expectedTxHashesAndRollups =
          ROLLUP_BLOB_TRANSACTION_INPUT.transactions.map(({ hash }) => [
            hash,
            expectedRollup,
          ]);

        expect(indexedTxHashesAndRollups).toEqual(expectedTxHashesAndRollups);
      });

      it("should categorize a rollup transaction correctly", async () => {
        await authorizedCaller.indexData(ROLLUP_BLOB_TRANSACTION_INPUT);

        const indexedTxHashesAndCategories =
          await authorizedCtx.prisma.transaction
            .findMany({
              select: {
                hash: true,
                from: {
                  select: {
                    rollup: true,
                  },
                },
              },
              where: {
                blockHash: ROLLUP_BLOB_TRANSACTION_INPUT.block.hash,
              },
            })
            .then((r) =>
              r.map((tx) => [tx.hash, tx.from.rollup ? "ROLLUP" : "OTHER"])
            );

        const expectedTxHashesAndCategories =
          ROLLUP_BLOB_TRANSACTION_INPUT.transactions.map(
            ({ hash }): [string, Category] => [hash, Category.ROLLUP]
          );
        expect(indexedTxHashesAndCategories).toEqual(
          expectedTxHashesAndCategories
        );
      });

      it("should categorize an unknown transaction correctly", async () => {
        const indexedTxHashesAndCategories =
          await authorizedCtx.prisma.transaction
            .findMany({
              select: {
                hash: true,
                from: {
                  select: {
                    rollup: true,
                  },
                },
              },
              where: {
                blockHash: INPUT.block.hash,
              },
              orderBy: [
                {
                  blockNumber: "asc",
                },
                {
                  index: "asc",
                },
              ],
            })
            .then((r) =>
              r.map((tx) => [tx.hash, tx.from.rollup ? "ROLLUP" : "OTHER"])
            );

        const expectedTxHashesAndCategories = INPUT.transactions.map(
          ({ hash }): [string, Category] => [hash, Category.OTHER]
        );

        expect(indexedTxHashesAndCategories).toEqual(
          expectedTxHashesAndCategories
        );
      });
    });

    describe("when indexing blobs", () => {
      it("should calculate and store blob size correctly", async () => {
        const expectedBlobSizes = INPUT.blobs.map((b) => ({
          versionedHash: b.versionedHash,
          size: b.data.slice(2).length / 2,
        }));

        const dbBlobSizes = await authorizedCtx.prisma.blob.findMany({
          select: {
            versionedHash: true,
            size: true,
          },
          where: {
            versionedHash: {
              in: expectedBlobSizes.map((b) => b.versionedHash),
            },
          },
        });

        expect(dbBlobSizes).toEqual(expectedBlobSizes);
      });

      it("should calculate and store blob usage size correctly", async () => {
        const expectedBlobUsageSizes = INPUT.blobs.map((b) => {
          let trailingZeroes = 0;
          let i = b.data.length - 1;

          while (i > 0 && b.data[i - 1] === "0") {
            trailingZeroes++;
            i--;
          }

          const totalBytes = b.data.slice(2).length / 2;
          const paddingBytes = trailingZeroes / 2;

          return {
            versionedHash: b.versionedHash,
            usageSize: totalBytes - paddingBytes,
          };
        });

        const dbBlobUsageSizes = await authorizedCtx.prisma.blob.findMany({
          select: {
            versionedHash: true,
            usageSize: true,
          },
          where: {
            versionedHash: {
              in: expectedBlobUsageSizes.map((b) => b.versionedHash),
            },
          },
        });

        expect(dbBlobUsageSizes).toEqual(expectedBlobUsageSizes);
      });

      it("should index them correctly", async () => {
        const txHashes = INPUT.transactions.map((tx) => tx.hash);
        const indexedBlobs = (
          await authorizedCtx.prisma.blobsOnTransactions.findMany({
            where: {
              txHash: {
                in: txHashes,
              },
            },
          })
        ).sort((a, b) => a.blobHash.localeCompare(b.blobHash));

        expect(indexedBlobs).toMatchInlineSnapshot(`
            [
              {
                "blobHash": "blobHash1000",
                "blockHash": "blockHash2010",
                "blockNumber": 2010,
                "blockTimestamp": 2023-09-01T13:50:21.000Z,
                "index": 0,
                "txHash": "txHash1000",
                "txIndex": 1,
              },
              {
                "blobHash": "blobHash1001",
                "blockHash": "blockHash2010",
                "blockNumber": 2010,
                "blockTimestamp": 2023-09-01T13:50:21.000Z,
                "index": 1,
                "txHash": "txHash1000",
                "txIndex": 1,
              },
              {
                "blobHash": "blobHash999",
                "blockHash": "blockHash2010",
                "blockNumber": 2010,
                "blockTimestamp": 2023-09-01T13:50:21.000Z,
                "index": 0,
                "txHash": "txHash999",
                "txIndex": 0,
              },
            ]
          `);
      });

      describe("when indexing blob data", () => {
        let ctxWithBlobPropagator: TRPCContext;
        let callerWithBlobPropagator: IndexerCaller;
        let blobPropagatorSpy: SpyInstance<
          [
            blobs: {
              versionedHash: string;
              data: string;
            }[]
          ],
          Promise<void>
        >;

        beforeAll(async () => {
          ctxWithBlobPropagator = await createTestContext({
            apiClient: "indexer",
            withBlobPropagator: true,
          });

          callerWithBlobPropagator = createIndexerCaller(ctxWithBlobPropagator);

          blobPropagatorSpy = vi.spyOn(
            // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
            ctxWithBlobPropagator.blobPropagator!,
            "propagateBlobs"
          );
        });

        it("should call blob propagator", async () => {
          const expectedInput = INPUT_WITH_DUPLICATED_BLOBS.blobs.map((b) => ({
            ...b,
            blockNumber: INPUT_WITH_DUPLICATED_BLOBS.block.number,
          }));

          await callerWithBlobPropagator.indexData(INPUT_WITH_DUPLICATED_BLOBS);

          expect(blobPropagatorSpy).toHaveBeenCalledOnce();
          expect(
            blobPropagatorSpy,
            "Propagator called with invalid blobs"
          ).toHaveBeenCalledWith(expectedInput);
        });
      });
    });

    it("should create valid blob on transaction entities", async () => {
      const indexedBlobsOnTransactions = (
        await authorizedCtx.prisma.blobsOnTransactions.findMany({
          where: {
            txHash: {
              in: INPUT.transactions.map((tx) => tx.hash),
            },
          },
        })
      ).sort((a, b) => a.blobHash.localeCompare(b.blobHash));

      expect(indexedBlobsOnTransactions).toMatchInlineSnapshot(`
          [
            {
              "blobHash": "blobHash1000",
              "blockHash": "blockHash2010",
              "blockNumber": 2010,
              "blockTimestamp": 2023-09-01T13:50:21.000Z,
              "index": 0,
              "txHash": "txHash1000",
              "txIndex": 1,
            },
            {
              "blobHash": "blobHash1001",
              "blockHash": "blockHash2010",
              "blockNumber": 2010,
              "blockTimestamp": 2023-09-01T13:50:21.000Z,
              "index": 1,
              "txHash": "txHash1000",
              "txIndex": 1,
            },
            {
              "blobHash": "blobHash999",
              "blockHash": "blockHash2010",
              "blockNumber": 2010,
              "blockTimestamp": 2023-09-01T13:50:21.000Z,
              "index": 0,
              "txHash": "txHash999",
              "txIndex": 0,
            },
          ]
        `);
    });

    it("should update address entities correctly", async () => {
      // Remove duplicates
      const addressesSet = new Set(
        INPUT.transactions.flatMap((tx) => [tx.from, tx.to])
      );
      const indexedAddresses = await authorizedCtx.prisma.address.findMany({
        select: {
          address: true,
          firstBlockNumberAsReceiver: true,
          firstBlockNumberAsSender: true,
        },
        where: {
          address: {
            in: Array.from(addressesSet),
          },
        },
        orderBy: [
          {
            address: "asc",
          },
        ],
      });

      expect(indexedAddresses).toMatchInlineSnapshot(`
          [
            {
              "address": "address10",
              "firstBlockNumberAsReceiver": 2010,
              "firstBlockNumberAsSender": null,
            },
            {
              "address": "address2",
              "firstBlockNumberAsReceiver": 1001,
              "firstBlockNumberAsSender": 1003,
            },
            {
              "address": "address7",
              "firstBlockNumberAsReceiver": null,
              "firstBlockNumberAsSender": 2010,
            },
            {
              "address": "address9",
              "firstBlockNumberAsReceiver": null,
              "firstBlockNumberAsSender": 2010,
            },
          ]
        `);
    });

    it("should be idempotent", async () => {
      // Index the same block again
      await expect(authorizedCaller.indexData(INPUT)).resolves.toBeUndefined();
    });

    testValidError(
      "should fail when no blob propagator is defined",
      async () => {
        await noPropagatorCaller.indexData(INPUT);
      },
      TRPCError,
      {
        checkCause: true,
      }
    );

    testValidError(
      "should fail when receiving an empty array of transactions",
      async () => {
        await authorizedCaller.indexData({
          ...INPUT,
          transactions: [] as unknown as typeof INPUT.transactions,
        });
      },
      TRPCError
    );

    testValidError(
      "should fail when receiving an empty array of blobs",
      async () => {
        await authorizedCaller.indexData({
          ...INPUT,
          blobs: [] as unknown as typeof INPUT.blobs,
        });
      },
      TRPCError
    );
  });

  testValidError(
    "should fail when calling procedure without auth",
    async () => {
      await nonAuthorizedCaller.indexData(INPUT);
    },
    TRPCError
  );
});
