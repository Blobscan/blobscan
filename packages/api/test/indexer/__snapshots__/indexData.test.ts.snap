// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`indexData > should fail when calling procedure without auth 1`] = `"UNAUTHORIZED"`;

exports[`indexData > when authorized > should fail when no blob propagator is defined 1`] = `"Blob propagator is missing"`;

exports[`indexData > when authorized > should fail when no blob propagator is defined 2`] = `undefined`;

exports[`indexData > when authorized > should fail when receiving an empty array of blobs 1`] = `
"[
  {
    \\"code\\": \\"too_small\\",
    \\"minimum\\": 1,
    \\"type\\": \\"array\\",
    \\"inclusive\\": true,
    \\"exact\\": false,
    \\"message\\": \\"Array must contain at least 1 element(s)\\",
    \\"path\\": [
      \\"blobs\\"
    ]
  }
]"
`;

exports[`indexData > when authorized > should fail when receiving an empty array of transactions 1`] = `
"[
  {
    \\"code\\": \\"too_small\\",
    \\"minimum\\": 1,
    \\"type\\": \\"array\\",
    \\"inclusive\\": true,
    \\"exact\\": false,
    \\"message\\": \\"Array must contain at least 1 element(s)\\",
    \\"path\\": [
      \\"transactions\\"
    ]
  }
]"
`;
